ccl/schemachangerccl: TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid failed

ccl/schemachangerccl.TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/11151110?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/11151110?buildTab=artifacts#/) on master @ [8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6](https://github.com/cockroachdb/cockroach/commits/8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 969923 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
        true
    datadriven.go:143: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1064/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/schemachangerccl/schemachangerccl_test_/schemachangerccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_primary_key_drop_rowid/alter_table_alter_primary_key_drop_rowid.definition:19:
        stage-exec [2 args]
        INSERT INTO t VALUES($stageKey);
        INSERT INTO t VALUES($stageKey + 1);
        ----
        pq: INSERT has more expressions than target columns, 2 expressions for 1 targets
    datadriven.go:143: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1064/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/schemachangerccl/schemachangerccl_test_/schemachangerccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_primary_key_drop_rowid/alter_table_alter_primary_key_drop_rowid.definition:27:
        stage-query [2 args]
        SELECT count(*)=$successfulStageCount*2 FROM t;
        ----
        true
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_1_of_15/restore_all_tables_in_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_3_of_15/restore_all_tables_in_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_3_of_15/restore_database_with_schema_only
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_4_of_15/restore_all_tables_in_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_4_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_6_of_15/restore_database_with_schema_only
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_8_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_8_of_15/restore_database_with_schema_only
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_10_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_1_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_1_of_15/restore_database_with_schema_only
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_5_of_15/restore_database_with_schema_only
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_7_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_7_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_8_of_15/restore_all_tables_in_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_10_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_6_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_9_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_3_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_6_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_9_of_15/restore_database_with_schema_only
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_3_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_1_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_2_of_15/restore_all_tables_in_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_7_of_15/restore_all_tables_in_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_8_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_10_of_15/restore_all_tables_in_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_2_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_4_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_4_of_15/restore_database_with_schema_only
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_5_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_5_of_15/restore_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_9_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_2_of_15
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_6_of_15/restore_all_tables_in_database
=== RUN   TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid/post_commit_stage_11_of_15
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupSuccessMixedVersion_base_alter_table_alter_primary_key_drop_rowid.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30283

--- a/pkg/sql/schemachanger/sctest/backup.go
+++ b/pkg/sql/schemachanger/sctest/backup.go
@@ -13,7 +13,9 @@ package sctest
 import (
 	"context"
 	gosql "database/sql"
+	"flag"
 	"fmt"
+	"math/rand"
 	"strings"
 	"sync/atomic"
 	"testing"
@@ -70,10 +72,6 @@ func BackupSuccessMixedVersion(t *testing.T, path string, factory TestServerFact
 	// and at least as expensive to run.
 	skip.UnderShort(t)
 
-	if strings.Contains(path, "alter_table_add_primary_key_drop_rowid") {
-		skip.WithIssue(t, 107552, "flaky test")
-	}
-
 	factory = factory.WithMixedVersion()
 	cumulativeTestForEachPostCommitStage(t, path, factory, func(t *testing.T, cs CumulativeTestCaseSpec) {
 		backupSuccess(t, factory, cs)
@@ -96,7 +94,27 @@ func BackupRollbacksMixedVersion(t *testing.T, path string, factory TestServerFa
 	})
 }
 
+// runAllBackups runs all the backup tests, disabling the random skipping.
+var runAllBackups = flag.Bool(
+	"run-all-backups", false,
+	"if true, run all backups instead of a random subset",
+)
+
+// If the number of stages in the same phase exceeds skipThreshold, we enable
+// skipping such that the backup test for each stage is skipped with probability
+// skipRate.
+// Set runAllBackups to true to disable skipping altogether.
+const skipThreshold = 10
+const skipRate = .5
+
+func maybeRandomlySkip(t *testing.T, stageCountInPhase int) {
+	if !*runAllBackups && stageCountInPhase > skipThreshold && rand.Float64() < skipRate {
+		skip.IgnoreLint(t, "skipping due to randomness")
+	}
+}
+
 func backupSuccess(t *testing.T, factory TestServerFactory, cs CumulativeTestCaseSpec) {
+	maybeRandomlySkip(t, cs.StagesCount)
 	ctx := context.Background()
 	url := fmt.Sprintf("userfile://backups.public.userfiles_$user/data_%s_%d",
 		cs.Phase, cs.StageOrdinal)
@@ -205,6 +223,7 @@ func backupRollbacks(t *testing.T, factory TestServerFactory, cs CumulativeTestC
 	if cs.Phase != scop.PostCommitPhase {
 		return
 	}
+	maybeRandomlySkip(t, cs.StagesCount)
 	ctx := context.Background()
 	var urls atomic.Value
 	var dbForBackup atomic.Pointer[gosql.DB]

--- a/pkg/sql/schemachanger/sctest/framework.go
+++ b/pkg/sql/schemachanger/sctest/framework.go
@@ -644,83 +644,90 @@ func (cs CumulativeTestCaseSpec) run(t *testing.T, fn func(t *testing.T)) bool {
 	return t.Run(fmt.Sprintf("%s_stage_%d_of_%d", prefix, cs.StageOrdinal, cs.StagesCount), fn)
 }
 
+// cumulativeTestForEachPostCommitStage invokes `tf` once for each stage in the
+// PostCommitPhase. These invocation are run in parallel.
 func cumulativeTestForEachPostCommitStage(
 	t *testing.T,
 	relTestCaseDir string,
 	factory TestServerFactory,
 	tf func(t *testing.T, spec CumulativeTestCaseSpec),
 ) {
-	testFunc := func(t *testing.T, spec CumulativeTestSpec) {
-		// Skip this test if any of the stmts is not fully supported.
-		if err := areStmtsFullySupportedAtClusterVersion(t, spec, factory); err != nil {
-			skip.IgnoreLint(t, "test is skipped because", err.Error())
-		}
-		var postCommitCount, postCommitNonRevertibleCount int
-		var after [][]string
-		var dbName string
-		prepfn := func(db *gosql.DB, p scplan.Plan) {
-			for _, s := range p.Stages {
-				switch s.Phase {
-				case scop.PostCommitPhase:
-					postCommitCount++
-				case scop.PostCommitNonRevertiblePhase:
-					postCommitNonRevertibleCount++
+	// Grouping the parallel subtests into a non-parallel subtest allows any defer
+	// calls to work as expected.
+	t.Run("group", func(t *testing.T) {
+		testFunc := func(t *testing.T, spec CumulativeTestSpec) {
+			// Skip this test if any of the stmts is not fully supported.
+			if err := areStmtsFullySupportedAtClusterVersion(t, spec, factory); err != nil {
+				skip.IgnoreLint(t, "test is skipped because", err.Error())
+			}
+			var postCommitCount, postCommitNonRevertibleCount int
+			var after [][]string
+			var dbName string
+			prepfn := func(db *gosql.DB, p scplan.Plan) {
+				for _, s := range p.Stages {
+					switch s.Phase {
+					case scop.PostCommitPhase:
+						postCommitCount++
+					case scop.PostCommitNonRevertiblePhase:
+						postCommitNonRevertibleCount++
+					}
+				}
+				tdb := sqlutils.MakeSQLRunner(db)
+				var ok bool
+				dbName, ok = maybeGetDatabaseForIDs(t, tdb, screl.AllTargetStateDescIDs(p.TargetState))
+				if ok {
+					tdb.Exec(t, fmt.Sprintf("USE %q", dbName))
+				}
+				after = tdb.QueryStr(t, fetchDescriptorStateQuery)
+			}
+			withPostCommitPlanAfterSchemaChange(t, spec, factory, prepfn)
+			if postCommitCount+postCommitNonRevertibleCount == 0 {
+				skip.IgnoreLint(t, "test case has no post-commit stages")
+				return
+			}
+			if dbName == "" {
+				skip.IgnoreLint(t, "test case has no usable database")
+				return
+			}
+			var testCases []CumulativeTestCaseSpec
+			for stageOrdinal := 1; stageOrdinal <= postCommitCount; stageOrdinal++ {
+				testCases = append(testCases, CumulativeTestCaseSpec{
+					CumulativeTestSpec: spec,
+					Phase:              scop.PostCommitPhase,
+					StageOrdinal:       stageOrdinal,
+					StagesCount:        postCommitCount,
+					After:              after,
+					DatabaseName:       dbName,
+				})
+			}
+			for stageOrdinal := 1; stageOrdinal <= postCommitNonRevertibleCount; stageOrdinal++ {
+				testCases = append(testCases, CumulativeTestCaseSpec{
+					CumulativeTestSpec: spec,
+					Phase:              scop.PostCommitNonRevertiblePhase,
+					StageOrdinal:       stageOrdinal,
+					StagesCount:        postCommitNonRevertibleCount,
+					After:              after,
+					DatabaseName:       dbName,
+				})
+			}
+			var hasFailed bool
+			for _, tc := range testCases {
+				fn := func(t *testing.T) {
+					t.Parallel() // SAFE FOR TESTING
+					tf(t, tc)
+				}
+				if hasFailed {
+					fn = func(t *testing.T) {
+						skip.IgnoreLint(t, "skipping test cases subsequent to earlier failure")
+					}
+				}
+				if !tc.run(t, fn) {
+					hasFailed = true
 				}
 			}
-			tdb := sqlutils.MakeSQLRunner(db)
-			var ok bool
-			dbName, ok = maybeGetDatabaseForIDs(t, tdb, screl.AllTargetStateDescIDs(p.TargetState))
-			if ok {
-				tdb.Exec(t, fmt.Sprintf("USE %q", dbName))
-			}
-			after = tdb.QueryStr(t, fetchDescriptorStateQuery)
 		}
-		withPostCommitPlanAfterSchemaChange(t, spec, factory, prepfn)
-		if postCommitCount+postCommitNonRevertibleCount == 0 {
-			skip.IgnoreLint(t, "test case has no post-commit stages")
-			return
-		}
-		if dbName == "" {
-			skip.IgnoreLint(t, "test case has no usable database")
-			return
-		}
-		var testCases []CumulativeTestCaseSpec
-		for stageOrdinal := 1; stageOrdinal <= postCommitCount; stageOrdinal++ {
-			testCases = append(testCases, CumulativeTestCaseSpec{
-				CumulativeTestSpec: spec,
-				Phase:              scop.PostCommitPhase,
-				StageOrdinal:       stageOrdinal,
-				StagesCount:        postCommitCount,
-				After:              after,
-				DatabaseName:       dbName,
-			})
-		}
-		for stageOrdinal := 1; stageOrdinal <= postCommitNonRevertibleCount; stageOrdinal++ {
-			testCases = append(testCases, CumulativeTestCaseSpec{
-				CumulativeTestSpec: spec,
-				Phase:              scop.PostCommitNonRevertiblePhase,
-				StageOrdinal:       stageOrdinal,
-				StagesCount:        postCommitNonRevertibleCount,
-				After:              after,
-				DatabaseName:       dbName,
-			})
-		}
-		var hasFailed bool
-		for _, tc := range testCases {
-			fn := func(t *testing.T) {
-				tf(t, tc)
-			}
-			if hasFailed {
-				fn = func(t *testing.T) {
-					skip.IgnoreLint(t, "skipping test cases subsequent to earlier failure")
-				}
-			}
-			if !tc.run(t, fn) {
-				hasFailed = true
-			}
-		}
-	}
-	cumulativeTest(t, relTestCaseDir, testFunc)
+		cumulativeTest(t, relTestCaseDir, testFunc)
+	})
 }
 
 // fetchDescriptorStateQuery returns the CREATE statements for all descriptors

